rfcs/rdm-0056-event-bus.md

```diff
@@ -0,0 +1,311 @@
+
+**To events definition**
+
+In this RFC we opted by defining the _action_ e.g. `RecordCreatedEvent` in the event class name, instead of having a generic `RecordEvent` with an `action` property. This was to ease the configuration and processing of the handlers.
+
+In addition, if events should be heavy or light was discussed. Opting by light events, since it would be easier to add more information than to remove it.
+
+_Heavy events_
+A heavy event would contain enough information for a handler to replicate the fact, avoiding that way extra expensive operations (e.g. db query, permission checks, etc.).
+
+```python
+@dataclass
+class RecordCreatedEvent(RecordEvent):
+    created: datetime
+    updated: datetime
+    recid: str
+    revision_id: int
+    metadata: dict
+    access: dict
+    ...
+```
+
+On the other hand, heavy events might suppose a challenge in memory terms to the queues. In addition, if they are sent to another system (e.g. Elasticsearch) for example for audit logging, they would require the implementation of post processing logic to anonymize.
+
+_Light events_
+A light event would contain only the minimum information needed for a handler to achieve it's goal, but it will require service calls (i.e. DB queries, permission checks, etc.).
+
+```python
+@dataclass
+class RecordCreatedEvent(RecordEvent):
+    recid: str
+```
+
+Light events might allow for easier composition, e.g. add request context information to a `RecordReadEvent` if it happend on the Web UI.
+
+**To configuration loading**
+
+Several options were taken into account to allow the customization of handlers. Both entrypoints and class configurations (e.g. like `ServiceConfig`) were discarded because of the complexity they add, while all that is looked for here is a simple way to establish a mapping between events and handlers (a _dict_).
+
+**To event worker task**
+
+Two alternatives were considered to the event worker task are:
+
+- Delegating the handling to the event bus itself. However, this would put even more responsability on `invenio-rdm-records`, which is not desired at the moment.
+- Having a new app run with for example a `supervisord`. This would increase the maintainance cost significantly to provide scalability, resilience, etc.
+
+## Unresolved questions
+
+**How to add events that require request context**
+
+Is this suitable for cases where request context is required? For example usage statistics would require IP addr, referer, User Agent among others. These are available at _resource_ level, how can an event be added/queued from there?
+
+**How to deal with cases that do not require a UoW**
+
+For example, record `read` operations do not require a _unit of work_ and therefore it is not possible add events such as `RecordReadEvent`.
+
+**Error handling**
+
+Error handling plays a bit role in this event bus implementation. While there are mechanisms to potentially allow persistance of events there are no mechanisms to:
```

Comment:
This is a very good point: the event success should be recorded only when the handler has completed the job. It is probably the part that should be expanded in this RFC.